src/rrdp.rs

```diff
@@ -1,456 +1,962 @@
+
+        let notification = self.make_notification_file()?;
+
+        let mut bytes: Vec<u8> = vec![];
+        let mut writer = rpki::xml::encode::Writer::new_with_indent(&mut bytes);
+        notification.to_xml(&mut writer)?;
+
+        file_ops::write_buf(&tmp_path, &bytes)
+            .with_context(|| "Could not write temporary notification file")?;
+
+        fs::rename(tmp_path, final_path)
+            .with_context(|| "Could not rename tmp notification file to real notification file")?;
+
+        Ok(())
+    }
+
+    fn make_notification_file(&self) -> Result<NotificationFile> {
+        let base_uri = self
+            .notification_uri
+            .parent()
+            .ok_or_else(|| anyhow!("Notification URI does not have a parent?!"))?;
+
+        let rel_path_snapshot = Self::rel_path_snapshot(self.session_id(), self.serial());
+
+        let snapshot_uri = base_uri
+            .join(rel_path_snapshot.as_bytes())
+            .with_context(|| "Could not derive snapshot URI")?;
+
+        let snapshot_hash = self.snapshot.hash();
+        let snapshot_info = SnapshotInfo::new(snapshot_uri, snapshot_hash);
+
+        let mut deltas = vec![];
+        for delta in &self.deltas {
+            let serial = delta.serial();
+            let hash = delta.hash();
+            let rel_path_delta = Self::rel_path_delta(self.session_id(), serial);
+            let uri = base_uri
+                .join(rel_path_delta.as_bytes())
+                .with_context(|| "Could not derive delta URI")?;
+
+            deltas.push(DeltaInfo::new(serial, uri, hash));
         }
+
+        Ok(NotificationFile::new(
+            self.session_id,
+            self.serial,
+            snapshot_info,
+            deltas,
+        ))
     }
-}
 
-impl ProcessDelta for DeltaParser<'_> {
-    type Err = DeltaParserError;
+    /// Marks a snapshot file as deprecated. Assumes that the session id is still unchanged. If the
+    /// there would be a session id reset, then deprecate files for the old session id first, before
+    /// updating the current session id.
+    fn deprecate_snapshot_file(&mut self) {
+        let path = Self::path_snapshot(&self.rrdp_dir, self.session_id(), self.serial());
+        self.deprecated_files.push(DeprecatedFile::new(path));
+    }
+
+    /// Writes a new notification file. Will not check whether the file already
+    /// exists because this is assumed to be called for new snapshot files only.
+    fn write_snapshot(&self) -> Result<()> {
+        let path = Self::path_snapshot(&self.rrdp_dir, self.session_id(), self.serial());
+        if path.exists() {
+            debug!("Skip writing existing snapshot file to {:?}", path)
+        } else {
+            info!("Writing snapshot file to {:?}", path);
+            let xml = self
+                .snapshot
+                .xml()
+                .ok_or_else(|| anyhow!("Snapshot XML not recovered on startup"))?;
```

Comment:
Which snapshot XML? Should the error context mention the path?

Reply:
done